### PR TITLE
ceph: add ldap configs for rgw

### LINKED ISSUE #4315: Support RGW LDAP integration
**Is this a bug report or feature request?**
* Feature Request

**What should the feature do:**
The Ceph Object Gateway supports integrating with LDAP for authenticating and creating users, see: https://docs.ceph.com/docs/master/radosgw/ldap-auth/
It would be nice if this was directly supported in the "object" part of the rook-ceph integration.
**What is use case behind this feature:**
Providing this feature would enable easy integration with a corporate LDAP server for use of the object store, including its S3 compatible interface.  This would prevent having to have someone create individual user accounts via the "object-user.yaml" for every user that needs access to the object store, and allow for control based on group membership(s) in the LDAP server.
**Environment**:

This can currently be mostly achieved using existing "escape hatches" in rook.  The turning on of the LDAP integration can be done by adding config values to the "rook-config-override" configmap similar to the following (actual values removed):
[global]
    rgw ldap binddn = 
    rgw_ldap_secret = /etc/ceph/ldap/bindpass.secret
    rgw ldap uri = 
    rgw ldap searchdn =
    rgw ldap dnattr =
    rgw_s3_auth_use_ldap = true

The main problem comes from the fact that "rgw_ldap_secret" needs to point to a file that contains the unencrypted/unencoded password for the account being used to query the LDAP server that is specified by "rgw_ldap_binddn".   Mounting this file into the "rook-ceph-rgw-my-store" related pod is difficult/fragile at the moment.  By creating a secret that contains the password, we can mount that secret as a volume in the pod, but that requires modifying/patching the "rook-ceph-rgw-my-store" deployment similar to below:
spec:
  template:
    spec:
      containers:
      - name: rgw
        volumeMounts:
        - mountPath: /etc/ceph/ldap
          name: rook-ceph-rgw-ldap-pass
          readOnly: true
      volumes:
      - name: rook-ceph-rgw-ldap-pass
        secret:
          defaultMode: 420
          secretName: rook-ceph-rgw-ldap-pass

This works, and the LDAP integration succeeds.  But it is fragile, in that anything that causes the operator to restart/redeploy the  "rook-ceph-rgw-my-store" deployment overwrites the above change, and it needs to be reapplied.

### PR DESCRIPTION
**Pull request by thotz**
The Ceph Object Gateway supports integrating with LDAP for authenticating
and creating users, for that required config options need to be added in
cephobjectstore CRD.

Resolves #4315
Signed-off-by: Jiffin Tony Thottan <thottanjiffin@gmail.com>
**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

File: pkg/operator/ceph/object/config.go [hunk 1]
```diff
@@ -44,6 +44,10 @@ caps osd = "allow rwx"
 	caBundleExtractedDir            = caBundleTrustedDir + "extracted/"
 	caBundleKeyName                 = "cabundle"
 	caBundleFileName                = "custom-ca-bundle.crt"
+	ldapVolumeName                  = "rook-ceph-rgw-ldap"
+	ldapDir                         = "/etc/ceph/ldap"
```
Comment (BlaineEXE) [inline pkg/operator/ceph/object/config.go:48]: How about a different location for these files? Documents I found suggest `/etc/` is often used, which we also shouldn't use. I think `/etc/ldap` would be betters.

File: cluster/examples/kubernetes/ceph/object-openshift.yaml [hunk 2]
```diff
@@ -129,6 +134,14 @@ spec:
...
 # data:
 #   token: ROOK_TOKEN_CHANGE_ME
+# ---
+# apiVersion: v1
+# kind: Secret
+# metadata:
+#   name: rook-ldap-rgw
+#   namespace: rook-ceph # namespace:cluster
+# data:
+#   secret: PASSWORD_CHANGE_ME
```
Comment (BlaineEXE) [inline cluster/examples/kubernetes/ceph/object-openshift.yaml:144]: If it's a password, we should just call it `password` for maximum clarity.

```suggestion
#   password: CHANGE_ME
```

**Review by BlaineEXE — CHANGES_REQUESTED** [review]
Let's design more explicitly for usability and developer understanding.

Also, I will again ask where the unit tests are for this feature.

We will want community help testing this feature, and it would be most helpful if someone could help us set up an integration test for this so that we can be sure we aren't breaking it in the future.

File: pkg/apis/ceph.rook.io/v1/types.go [hunk 3]
```diff
@@ -1277,6 +1277,36 @@ type ObjectStoreSpec struct {
 	// +optional
 	// +nullable
 	Security *SecuritySpec `json:"security,omitempty"`
+
+	// LDAP represents ldap settings
+	// +optional
+	// +nullable
+	LDAP *LDAPSpec `json:"ldap,omitempty"`
```
Comment (leseb) [inline pkg/apis/ceph.rook.io/v1/types.go:1284]: I think this needs to be in a higher level Spec. LDAP and OPA for instance could fall under the same category. Perhaps something `'IdentityManagement`?

File: pkg/apis/ceph.rook.io/v1/types.go [hunk 4]
```diff
@@ -1277,6 +1277,36 @@ type ObjectStoreSpec struct {
...
+}
+
+type LDAPSpec struct {
```
Comment (leseb) [inline pkg/apis/ceph.rook.io/v1/types.go:1287]: `LDAPSpec` needs a comment

**Review by leseb — CHANGES_REQUESTED** [review]
We also need unit tests for the spec generation, thanks!

**Comment by travisn** [comment]
@thotz There are a number of old comments and questions on this PR, what is the status? According to the discussion in #4315, there is still interest in completing this work. Thanks!

**Comment by thotz** [comment]
> @thotz There are a number of old comments and questions on this PR, what is the status? According to the discussion in #4315, there is still interest in completing this work. Thanks!

One of the ideas was from @leseb to merge this with [opa feature](https://github.com/rook/rook/pull/8731), I was not intended to test both features and it kinda blocked me from further development. I will refresh this PR by addressing existing commands

**Comment by cccsss01** [comment]
I am interested in this as well!

**Comment by travisn** [comment]
Closing in favor of #11091

**Pull request closed without merging**